File: python/src/kickstarter/json_flight_consolidator_app.py
import os
from pyflink.common import Configuration, WatermarkStrategy
from pyflink.datastream import StreamExecutionEnvironment, DataStream
from pyflink.datastream.connectors.kafka import KafkaSource, KafkaSink, KafkaRecordSerializationSchema, KafkaOffsetsInitializer, DeliveryGuarantee
from pyflink.datastream.formats.json import JsonRowDeserializationSchema, JsonRowSerializationSchema
from pyflink.datastream.checkpoint_config import ExternalizedCheckpointRetention
from pyflink.table import StreamTableEnvironment
from pyflink.table.catalog import ObjectPath
from datetime import datetime, timezone
import logging

from model.flight_data import FlightData
from model.airline_flight_data import AirlineFlightData
from helper.common import load_catalog, load_database, get_confluent_properties


__copyright__  = "Copyright (c) 2024-2025 Jeffrey Jonathan Jennings"
__credits__    = ["Jeffrey Jonathan Jennings"]
__license__    = "MIT"
__maintainer__ = "Jeffrey Jonathan Jennings"
__email__      = "j3@thej3.com"
__status__     = "dev"


# Setup the logger
logger = logging.getLogger('FlyerStatsApp')

def main():
    """The entry point to the Flight Importer Flink App (a.k.a., Flink job graph --- DAG)."""
    # --- Retrieve environment variables
    service_account_user = os.getenv("SERVICE_ACCOUNT_USER", "")
    s3_bucket_name = os.getenv("AWS_S3_BUCKET", "")
    aws_region = os.getenv("AWS_REGION", "")

    # Get the Kafka Client properties from AWS Secrets Manager and AWS Systems Manager Parameter Store.
    secret_path_prefix = f"confluent_cloud_resource/{service_account_user}"

    """
    Using the `java_client` configuration instead of the `python_client` configuration
    because PyFlink converts into Java code and the Java code is what is executed.
    """
    consumer_properties, error_message = get_confluent_properties(
        aws_region_name=aws_region,
        kafka_cluster_secrets_path=f"{secret_path_prefix}/kafka_cluster/java_client",
        client_parameters_path=f"/{secret_path_prefix}/consumer_kafka_client"
    )
    if error_message:
        raise RuntimeError(f"Failed to retrieve the Confluent Cloud properties for the Kafka Consumer client because {error_message}")
    producer_properties, error_message = get_confluent_properties(
        aws_region_name=aws_region,
        kafka_cluster_secrets_path=f"{secret_path_prefix}/kafka_cluster/java_client",
        client_parameters_path=f"/{secret_path_prefix}/producer_kafka_client"
    )
    if error_message:
        raise RuntimeError(f"Failed to retrieve the Confluent Cloud properties for the Kafka Producer client because {error_message}")
    producer_properties['compression.type'] = 'lz4'

    # --- Create a configuration to force Avro serialization instead of Kyro serialization.
    config = Configuration()
    config.set_string("pipeline.force-avro", "true")

    # --- Configure parent-first classloading for metrics
    config.set_string("classloader.parent-first-patterns.additional", 
                    "com.codahale.metrics;io.dropwizard.metrics")

    # --- Create a blank Flink execution environment (a.k.a. the Flink job graph -- the DAG).
    env = StreamExecutionEnvironment.get_execution_environment(config)

    """
    Enable checkpointing every 10,000 milliseconds (10 seconds).  Note, consider the
    resource cost of checkpointing frequency, as short intervals can lead to higher
    I/O and CPU overhead.  Proper tuning of checkpoint intervals depends on the
    state size, latency requirements, and resource constraints.
    """
    env.enable_checkpointing(10000)

    # --- Set minimum pause between checkpoints to 5,000 milliseconds (5 seconds).
    env.get_checkpoint_config().set_min_pause_between_checkpoints(5000)

    # --- Set tolerable checkpoint failure number to 3.
    env.get_checkpoint_config().set_tolerable_checkpoint_failure_number(3)

    """
    Externalized Checkpoint Retention: RETAIN_ON_CANCELLATION" means that the system will keep the
    checkpoint data in persistent storage even if the job is manually canceled. This allows you to
    later restore the job from that last saved state, which is different from the default behavior,
    where checkpoints are deleted on cancellation. This setting requires you to manually clean up
    the checkpoint state later if it's no longer needed. 
    """
    env.get_checkpoint_config().set_externalized_checkpoint_retention(
        ExternalizedCheckpointRetention.RETAIN_ON_CANCELLATION
    )

    """
    Set checkpoint timeout to 60 seconds, which is the maximum amount of time a
    checkpoint attempt can take before being discarded.  Note, setting an appropriate
    checkpoint timeout helps maintain a balance between achieving exactly-once semantics
    and avoiding excessive delays that can impact real-time stream processing performance.
    """
    env.get_checkpoint_config().set_checkpoint_timeout(60000)

    """
    Set the maximum number of concurrent checkpoints to 1 (i.e., only one checkpoint
    is created at a time).  Note, this is useful for limiting resource usage and
    ensuring checkpoints do not interfere with each other, but may impact throughput
    if checkpointing is slow.  Adjust this setting based on the nature of your job,
    the size of the state, and available resources. If your environment has enough
    resources and you want to ensure faster recovery, you could increase the limit
    to allow multiple concurrent checkpoints.
    """
    env.get_checkpoint_config().set_max_concurrent_checkpoints(1)

    # Create a Table Environment
    tbl_env = StreamTableEnvironment.create(stream_execution_environment=env)

    # Sets up a Flink Kafka source to consume data from the Kafka topic `airline.skyone`
    skyone_source = (KafkaSource.builder()
                                .set_properties(consumer_properties)
                                .set_topics("airline.skyone")
                                .set_group_id("skyone_group")
                                .set_starting_offsets(KafkaOffsetsInitializer.earliest())
                                .set_value_only_deserializer(JsonRowDeserializationSchema
                                                             .builder()
                                                             .type_info(AirlineFlightData.get_value_type_info())
                                                             .build())
                                .build())

    # Takes the results of the Kafka source and attaches the unbounded data stream
    skyone_stream = (env.from_source(skyone_source, WatermarkStrategy.no_watermarks(), "skyone_source")
                        .uid("skyone_source"))

    # Sets up a Flink Kafka source to consume data from the Kafka topic `airline.sunset`
    sunset_source = (KafkaSource.builder()
                                .set_properties(consumer_properties)
                                .set_topics("airline.sunset")
                                .set_group_id("sunset_group")
                                .set_starting_offsets(KafkaOffsetsInitializer.earliest())
                                .set_value_only_deserializer(JsonRowDeserializationSchema
                                                             .builder()
                                                             .type_info(AirlineFlightData.get_value_type_info())
                                                             .build())
                                .build())

    # Takes the results of the Kafka source and attaches the unbounded data stream
    sunset_stream = (env.from_source(sunset_source, WatermarkStrategy.no_watermarks(), "sunset_source")
                        .uid("sunset_source"))

    # Initialize the KafkaSink builder
    kafka_sink_builder = KafkaSink.builder().set_bootstrap_servers(producer_properties['bootstrap.servers'])

    # Loop through the producer properties and set each property
    for key, value in producer_properties.items():
        if key != 'bootstrap.servers':  # Skip the bootstrap.servers as it is already set
            kafka_sink_builder.set_property(key, value)

    # Sets up a Flink Kafka sink to produce data to the Kafka topic `airline.flyer_stats`
    flight_sink = (kafka_sink_builder                            
                   .set_record_serializer(KafkaRecordSerializationSchema
                                          .builder()
                                          .set_topic("airline.flight")
                                          .set_value_serialization_schema(JsonRowSerializationSchema
                                                                          .builder()
                                                                          .with_type_info(FlightData.get_value_type_info())
                                                                          .build())
                                          .build())
                    .set_delivery_guarantee(DeliveryGuarantee.EXACTLY_ONCE)
                    .set_transactional_id_prefix("json-flight-data-")  # apply unique prefix to prevent backchannel conflicts and potential memory leaks
                    .build())
    
    # --- Load Apache Iceberg catalog
    catalog = load_catalog(tbl_env, aws_region, s3_bucket_name, "apache_kickstarter")

    logging.info("Current catalog: %s", tbl_env.get_current_catalog())

    # --- Load database
    load_database(tbl_env, catalog, "airlines")

    logging.info("Current database: %s", tbl_env.get_current_database())

    """
    An ObjectPath in Apache Flink is a class that represents the fully qualified path to a
    catalog object, such as a table, view, or function.  It uniquely identifies an object
    within a catalog by encapsulating both the database name and the object name.  For
    instance, this case we using it to get the fully qualified path of the `flight`
    table
    """
    flight_table_path = ObjectPath(tbl_env.get_current_database(), "flight")

    logging.info("Current table: %s", flight_table_path.get_full_name())

    # Check if the table exists.  If not, create it
    try:
        if not catalog.table_exists(flight_table_path):
            # Define the table using Flink SQL
            tbl_env.execute_sql(f"""
                CREATE TABLE {flight_table_path.get_full_name()} (
                    email_address STRING,
                    departure_time STRING,
                    departure_airport_code STRING,
                    arrival_time STRING,
                    arrival_airport_code STRING,
                    flight_number STRING,
                    confirmation STRING,
                    airline STRING
                ) PARTITIONED BY (arrival_airport_code)
                WITH (
                    'write.format.default' = 'parquet',
                    'write.target-file-size-bytes' = '134217728',
                    'write.delete.mode' = 'merge-on-read',
                    'write.update.mode' = 'merge-on-read',
                    'format-version' = '2'
                )
            """)

    except Exception as e:
        logging.error("A critical error occurred to during the processing of the table because %s", e)
        exit(1)

    # Combine the Airline DataStreams into one DataStream
    flight_datastream = combine_datastreams(skyone_stream, sunset_stream).map(lambda d: d.to_row(), output_type=FlightData.get_value_type_info())

    # Populate the Apache Iceberg Table with the data from the data stream
    (tbl_env.from_data_stream(flight_datastream)
            .execute_insert(flight_table_path.get_full_name()))

    # Sinks the Flight DataStream into a single Kafka topic
    flight_datastream.sink_to(flight_sink).name("json_flight_data_sink").uid("json_flight_data_sink")
    
    # Execute the Flink job graph (DAG)
    try:
        env.execute("json_flight_consolidator_app")
    except Exception as e:
        logging.error("The App stopped early due to the following: %s", e)

    # Combine the Airline DataStreams into one DataStream
    flight_datastream = combine_datastreams(skyone_stream, sunset_stream).map(lambda d: d.to_row(), output_type=FlightData.get_value_type_info())

    # Populate the Apache Iceberg Table with the data from the data stream
    (tbl_env.from_data_stream(flight_datastream)
            .execute_insert(flight_table_path.get_full_name()))

    # Sinks the Flight DataStream into a single Kafka topic
    (flight_datastream.sink_to(flight_sink)
                      .name("flight_sink")
                      .uid("flight_sink"))
    
    # Execute the Flink job graph (DAG)
    try:
        env.execute("json_flight_consolidator_app")
    except Exception as e:
        logging.error("The App stopped early due to the following: %s", e)


def combine_datastreams(skyone_stream: DataStream, sunset_stream: DataStream) -> DataStream:
    """This function combines the SkyOne Airlines and Sunset Air flight data streams into
    a single data stream.

    Args:
        skyone_stream (DataStream): is the source of the SkyOne Airlines flight data.
        sunset_stream (DataStream): is the source of the Sunset Air flight data.

    Returns:
        DataStream: the union of the SkyOne Airlines and Sunset Air flight data streams.
    """
    from helper.common import parse_isoformat

    # Map the data streams to the FlightData model and filter out Skyone flights that have already arrived
    skyone_flight_stream = (skyone_stream
                            .map(lambda flight: AirlineFlightData.to_flight_data("SkyOne", flight))
                            .name("map_skyone_to_flight_data")
                            .uid("map_skyone_to_flight_data")
                            .filter(lambda flight: parse_isoformat(flight.arrival_time) > datetime.now(timezone.utc))
                            .name("filter_skyone_future_flights")
                            .uid("filter_skyone_future_flights"))
    

    # Map the data streams to the FlightData model and filter out Sunset flights that have already arrived
    sunset_flight_stream = (sunset_stream
                            .map(lambda flight: AirlineFlightData.to_flight_data("Sunset", flight))
                            .name("map_sunset_to_flight_data")
                            .uid("map_sunset_to_flight_data")
                            .filter(lambda flight: parse_isoformat(flight.arrival_time) > datetime.now(timezone.utc))
                            .name("filter_sunset_future_flights")
                            .uid("filter_sunset_future_flights"))
    
    # Return the union of the two data streams
    return skyone_flight_stream.union(sunset_flight_stream)


if __name__ == "__main__":
    main()

File: java/app/src/main/java/kickstarter/JsonFlightConsolidatorApp.java
/**
 * Copyright (c) 2024-2025 Jeffrey Jonathan Jennings
 * 
 * @author Jeffrey Jonathan Jennings (J3)
 * 
 * 
 * This class imports flight data from `sunset` and `skyone` Kafka topics
 * and converts it to a unified format for the `flight` Kafka topic.
 * 
 * ------------------------------------------------------------------------------------------
 * I had a question, can you combine the Flink DataStream API and Table API in the same DAG?
 * 
 * The answer is yes, you can combine the Flink DataStream API and Table API in the same 
 * Directed Acyclic Graph (DAG) to leverage the strengths of both APIs within a single Flink
 * application.  This is particularly useful when you want to perform some complex event 
 * processing that is more naturally expressed in the DataStream API, and then switch to the 
 * Table API for more declarative and SQL-like processing, or vice versa.  This where I bring
 * Apache Iceberg into the mix.  Apache Iceberg is a table format that is designed to be used
 * with the Flink Table API.  In this DAG, I use Apache Iceberg to store the unified flight
 * data in the `db_example.airline_flight_data` table.
 */
package kickstarter;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Properties;

import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.configuration.ExternalizedCheckpointRetention;
import org.apache.flink.connector.base.DeliveryGuarantee;
import org.apache.flink.connector.kafka.sink.KafkaRecordSerializationSchema;
import org.apache.flink.connector.kafka.sink.KafkaSink;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
import org.apache.flink.formats.json.JsonSerializationSchema;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import kickstarter.helper.SnakeCaseJsonDeserializationSchema;
import kickstarter.model.AirlineJsonData;
import kickstarter.model.FlightJsonData;


public class JsonFlightConsolidatorApp {
    private static final Logger LOGGER = LoggerFactory.getLogger(JsonFlightConsolidatorApp.class);

    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
    

	/**
	 * The main method in a Flink application serves as the entry point of the program, where
	 * the Flink DAG is defined.  That is, the execution environment, the creation of the data
	 * streams or datasets, apply transformations, and trigger the execution of the application (by
	 * sending it to the Flink JobManager).
	 * 
	 * @param args list of strings passed to the main method from the command line.
	 * @throws Exception - The exceptions are forwarded, and are caught by the runtime.  
     * When the runtime catches an exception, it aborts the task and lets the fail-over logic
	 * decide whether to retry the task execution.
	 */
    public static void main(String[] args) throws Exception {
        /*
         * Retrieve the value(s) from the command line argument(s)
         */
        String serviceAccountUser = System.getenv().getOrDefault("SERVICE_ACCOUNT_USER", Common.getAppArgumentValue(args, Common.ARG_SERVICE_ACCOUNT_USER));

        // --- Validate required configuration
        if (serviceAccountUser == null || serviceAccountUser.trim().isEmpty()) {
            throw new IllegalArgumentException(
                "Service Account User is required. Set SERVICE_ACCOUNT_USER environment variable or use --service-account-user argument."
            );
        }

        LOGGER.info("Configuration loaded - Service Account User: {}", serviceAccountUser);

        // --- Create a blank Flink execution environment (a.k.a. the Flink job graph -- the DAG)
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        /*
         * Enable checkpointing every 10,000 milliseconds (10 seconds).  Note, consider the
         * resource cost of checkpointing frequency, as short intervals can lead to higher
         * I/O and CPU overhead.  Proper tuning of checkpoint intervals depends on the
         * state size, latency requirements, and resource constraints.
         */
        env.enableCheckpointing(10000);

        // --- Set minimum pause between checkpoints to 5,000 milliseconds (5 seconds).
        env.getCheckpointConfig().setMinPauseBetweenCheckpoints(5000);

        // --- Set tolerable checkpoint failure number to 3.
        env.getCheckpointConfig().setTolerableCheckpointFailureNumber(3);

        /*
         * Externalized Checkpoint Retention: RETAIN_ON_CANCELLATION" means that the system will keep the
         * checkpoint data in persistent storage even if the job is manually canceled. This allows you to
         * later restore the job from that last saved state, which is different from the default behavior,
         * where checkpoints are deleted on cancellation. This setting requires you to manually clean up
         * the checkpoint state later if it's no longer needed. 
         */
        env.getCheckpointConfig().setExternalizedCheckpointRetention(ExternalizedCheckpointRetention.RETAIN_ON_CANCELLATION);

        /*
         * Set checkpoint timeout to 60 seconds, which is the maximum amount of time a
         * checkpoint attempt can take before being discarded.  Note, setting an appropriate
         * checkpoint timeout helps maintain a balance between achieving exactly-once semantics
         * and avoiding excessive delays that can impact real-time stream processing performance.
         */
        env.getCheckpointConfig().setCheckpointTimeout(60000);

        /*
         * Set the maximum number of concurrent checkpoints to 1 (i.e., only one checkpoint
         * is created at a time).  Note, this is useful for limiting resource usage and
         * ensuring checkpoints do not interfere with each other, but may impact throughput
         * if checkpointing is slow.  Adjust this setting based on the nature of your job,
         * the size of the state, and available resources. If your environment has enough
         * resources and you want to ensure faster recovery, you could increase the limit
         * to allow multiple concurrent checkpoints.
         */
        env.getCheckpointConfig().setMaxConcurrentCheckpoints(1);
        
        // --- Kafka Consumer and Producer Client Properties
        Properties consumerProperties = Common.collectConfluentProperties(env, serviceAccountUser, true);
        Properties producerProperties = Common.collectConfluentProperties(env, serviceAccountUser, false);
        
        /*
         * Sets up a Flink Kafka source to consume data from the Kafka topic `skyone`
         */
        KafkaSource<AirlineJsonData> skyOneSource = KafkaSource.<AirlineJsonData>builder()
            .setProperties(consumerProperties)
            .setTopics("skyone")
            .setGroupId("skyone_group")
            .setStartingOffsets(OffsetsInitializer.earliest())
            .setValueOnlyDeserializer((new SnakeCaseJsonDeserializationSchema<>(AirlineJsonData.class)))
            .build();

        /*
         * Takes the results of the Kafka source and attaches the unbounded data stream to the Flink
         * environment (a.k.a. the Flink job graph -- the DAG)
         */
        DataStream<AirlineJsonData> skyOneStream = env
            .fromSource(skyOneSource, WatermarkStrategy.noWatermarks(), "skyone_source");

        /*
         * Sets up a Flink Kafka source to consume data from the Kafka topic `sunset`
         */
        KafkaSource<AirlineJsonData> sunsetSource = KafkaSource.<AirlineJsonData>builder()
            .setProperties(consumerProperties)
            .setTopics("sunset")
            .setGroupId("sunset_group")
            .setStartingOffsets(OffsetsInitializer.earliest())
            .setValueOnlyDeserializer(new SnakeCaseJsonDeserializationSchema<>(AirlineJsonData.class))
            .build();

        /*
         * Takes the results of the Kafka source and attaches the unbounded data stream to the Flink
         * environment (a.k.a. the Flink job graph -- the DAG)
         */
        DataStream<AirlineJsonData> sunsetStream = env
            .fromSource(sunsetSource, WatermarkStrategy.noWatermarks(), "sunset_source");

        /*
         * Sets up a Flink Kafka sink to produce data to the Kafka topic `flight` with the
         * specified serializer
         */
		KafkaRecordSerializationSchema<FlightJsonData> flightSerializer = KafkaRecordSerializationSchema.<FlightJsonData>builder()
            .setTopic("flight")
			.setValueSerializationSchema(new JsonSerializationSchema<>(Common::getMapper))
            .build();

        /*
         * Takes the results of the Kafka sink and attaches the unbounded data stream to the Flink
         * environment (a.k.a. the Flink job graph -- the DAG)
         */
        producerProperties.put("transaction.timeout.ms", "900000"); // 15m for long checkpoints
        producerProperties.put("enable.idempotence", "true");       // typically implied, explicit is fine
        KafkaSink<FlightJsonData> flightSink = KafkaSink.<FlightJsonData>builder()
            .setKafkaProducerConfig(producerProperties)
            .setTransactionalIdPrefix("json-flight-data-") // apply unique prefix to prevent backchannel conflicts and potential memory leaks
            .setDeliveryGuarantee(DeliveryGuarantee.EXACTLY_ONCE)
            .setRecordSerializer(flightSerializer)
            .build();

        /*
         * Defines the workflow for the Flink job graph (DAG) by connecting the data streams and
         * applying transformations to the data streams
         */
        defineWorkflow(skyOneStream, sunsetStream)
            .sinkTo(flightSink)
            .name("flightdata_sink");

        try {
            // --- Execute the Flink job graph (DAG)
            env.execute("JsonFlightConsolidatorApp");
        } catch (Exception e) {
            LOGGER.error("The App stopped early due to the following: ", e);
            throw e;
        }
    }

    /**
     * This method defines the workflow for the Flink job graph (DAG) by connecting the 
     * data streams and applying transformations to the data streams.
     * 
     * @param skyOneSource - The data stream source for the `skyone` Kafka topic
     * @param sunsetSource - The data stream source for the `sunset` Kafka topic
     * @return The data stream that is the result of the transformations
     */
	public static DataStream<FlightJsonData> defineWorkflow(DataStream<AirlineJsonData> skyOneSource, DataStream<AirlineJsonData> sunsetSource) {
        DataStream<FlightJsonData> skyOneFlightStream = 
            skyOneSource
                .filter(flight -> isValidFutureFlight(flight))
                .name("skyone_flight_filter").uid("skyone_flight_filter")
                .map(flight -> flight.toFlightData("SkyOne"))
                .name("skyone_flight_transform").uid("skyone_flight_transform");

		DataStream<FlightJsonData> sunsetFlightStream = 
            sunsetSource
                .filter(flight -> isValidFutureFlight(flight))
                .name("sunset_flight_filter").uid("sunset_flight_filter")
                .map(flight -> flight.toFlightData("Sunset"))
                .name("sunset_flight_transform").uid("sunset_flight_transform");

		return skyOneFlightStream.union(sunsetFlightStream);
    }

    /**
     * Validates that a flight has a valid arrival time in the future.
     * Includes null safety and exception handling to prevent job failures.
     * 
     * @param flight The flight data to validate
     * @return true if the flight has a valid future arrival time, false otherwise
     */
    private static boolean isValidFutureFlight(AirlineJsonData flight) {
        if (flight == null) {
            LOGGER.warn("Flight is null, filtering out");
            return false;
        }

        if (flight.getArrivalTime() == null) {
            LOGGER.warn("Flight has null arrival time, filtering out");
            return false;
        }

        try {
            LocalDateTime arrivalTime = LocalDateTime.parse(flight.getArrivalTime(), DATE_TIME_FORMATTER);
            return arrivalTime.isAfter(LocalDateTime.now());
        } catch (DateTimeParseException e) {
            LOGGER.warn("Failed to parse arrival time '{}': {}", flight.getArrivalTime(), e.getMessage());
            return false;
        }
    }
}